### ISSUE TITLE
Using strategy set on EmbedMany seems to call an unset then set

### ISSUE THREAD

**Issue by andrewjshults** [doctrine/mongodb-odm#434]
Been trying to debug some really weird race condition issues that happen when a bunch of AJAX actions are fired off. The AJAX call passes in a new array of recipients to add onto an RFP object. We setup a logger on the Doctrine mongo connection. The expected action is that it would simply call _set_, what seems to happen is that it first calls _unset_ on the field and then call _set_ with the correct data. Since mongo isn't transactional, this opens up the (very real) possibility that another thread will read the data in between the _unset_ and _set_ calls (where the data is missing). The default pushAll strategy does the same - the issue remains the same, but removing the unset would break the expected behavior (even doing a diff and setting/unsetting array elements would be subject to issues where the incorrect elements could be unset if the document had been updated in between the read and flush events).

From looking at the code, the set strategy seems to be the only one that would be safe from a race condition (although, we are still seeing multiple set queries for the same object within the same flush), as long as the unset is dropped. Trying to understand the logic behind including it to make sure that we're not going to break something else by removing it.

Also, the set and pushAll strategies seem to save the EmbedMany ArrayCollections in different formats - pushAll as an array and set as a hash (with integers stored as strings for keys). We're currently using the default pushAll strategy but plan to convert to the set strategy to (partially) solve this race condition.

```php
Array
(
    [update] => 1
    [query] => Array
        (
            [_id] => MongoId Object
                (
                    [$id] => 5080abcfd8ad73b13a000000
                )

        )

    [newObj] => Array
        (
            [$unset] => Array
                (
                    [recipients] => 1
                )

        )

    [options] => Array
        (
            [safe] => 1
            [fsync] => 1
        )

    [db] => test
    [collection] => rfps
)
```

```php
Array
(
    [update] => 1
    [query] => Array
        (
            [_id] => MongoId Object
                (
                    [$id] => 5080abcfd8ad73b13a000000
                )

        )

    [newObj] => Array
        (
            [$set] => Array
                (
                    [recipients] => Array
                        (
                            [0] => Array
                                (
                                    [name] =>  
                                    [email] => jeff@smith.com
                                    [contact] => Array
                                        (
                                            [$ref] => contacts
                                            [$id] => MongoId Object
                                                (
                                                    [$id] => 507ebffb4cd844af6e6873bc
                                                )

                                            [$db] => adstruc
                                        )

                                    [created] => MongoDate Object
                                        (
                                            [sec] => 1350611713
                                            [usec] => 0
                                        )

                                )

                            [1] => Array
                                (
                                    [name] =>  
                                    [email] => bob@bobdole.net
                                    [contact] => Array
                                        (
                                            [$ref] => contacts
                                            [$id] => MongoId Object
                                                (
                                                    [$id] => 507ebffb4cd844af6e6873bd
                                                )

                                            [$db] => adstruc
                                        )

                                    [created] => MongoDate Object
                                        (
                                            [sec] => 1350611713
                                            [usec] => 0
                                        )

                                )

                        )

                )

        )

    [options] => Array
        (
            [safe] => 1
            [fsync] => 1
        )

    [db] => adstruc
    [collection] => rfps
)
```

**Comment by malarzm** [comment]
Just to cross reference the issue will be fixed once #1120 is merged, also to help highly concurrent applications we have introduced `atomicSet` and `atomicSetArray` startegies for collections :)

**Comment by malarzm** [comment]
#1120 was merged manually in https://github.com/doctrine/mongodb-odm/commit/c06e8d40a3fd77a9c24481247890863ea858355c
